src/components/PopupWithForm.js

```diff
@@ -14,26 +24,42 @@ class PopupWithForm extends Popup {
     inputs.forEach((input) => {
       values[input.name] = input.value;
     });
+
     return values;
   }
 
   _submitForm = () => {
     const inputValues = this._getInputValues();
-    this._handleFormSubmit(inputValues);
+    this._loadingState = true;
+    this.showLoading(true);
+
+    setTimeout(() => {
+      this._handleFormSubmit(inputValues);
+
+      this._loadingState = false;
+      this.hideLoading();
+    }, 2000);
   };
 
+  showLoading(isLoading) {
+    if (isLoading) {
+      this._submitButton.textContent = "Saving...";
+    } else {
+      this._submitButton.textContent = "Save";
```

Comment:
You should return the initial button text here

One of the options to implement that would be saving it in the constructor and adding a parameter `loadingButtonText` to the constructor of `PopupWithForm` and using it to change button text. 

```js
constructor(popupSelector, { handleSubmit, loadingButtonText }) { 
  ...
  this._submitButton = this._formElement.querySelector(...)
  this._buttonText = this._submitButton.textContent; // save the initial button text
  this._loadingButtonText = loadingButtonText;
  …
}

showLoading() {
  this._submitButton.textContent = this._loadingButtonText
}

hideLoading() {
  this._submitButton.textContent = this._buttonText
}
``` 

The api call should look like this:

```js
popup.showLoading()
api.
  .then(...)
  .catch(...)
  .finally(() => {
    popup.hideLoading()

  })
```

Reply:
Needs correction: You should return the initial button text here (see the snipped above)

Reply:
You should return the initial button text here

One of the options to fix that would be saving the text in the constructor

```js
constructor(popupSelector, { handleSubmit }) { 
  …
  this._buttonText = buttonText = this._submitButton.textContent;
}

...
this._submitButton.textContent = this._buttonText
```